File: src/main/java/io/github/xiefrish2021/tag/compound/NBTElement.java
package io.github.xiefrish2021.tag.compound;

import io.github.xiefrish2021.NBTException;
import io.github.xiefrish2021.tag.array.ByteArrayTag;
import io.github.xiefrish2021.tag.array.IntArrayTag;
import io.github.xiefrish2021.tag.array.LongArrayTag;
import io.github.xiefrish2021.tag.list.ListTag;
import io.github.xiefrish2021.tag.*;
import io.github.xiefrish2021.ITag;

public record NBTElement(ITag tag) {
    @SuppressWarnings({"unchecked", "unused"})
    public <V extends ITag> ListTag<V> getAsList( Class<V> type) {
        try {
            return (ListTag<V>) tag;
        } catch (ClassCastException exception) {
            throw new NBTException("Value is not a list");
        }
    }

    public LongArrayTag getAsLongArray() {
        try {
            return ((LongArrayTag) tag);
        } catch (ClassCastException exception) {
            throw new NBTException("Value is not a long array");
        }
    }

    public ByteArrayTag getAsByteArray() {
        try {
            return ((ByteArrayTag) tag);
        } catch (ClassCastException exception) {
            throw new NBTException("Value is not a byte array");
        }
    }

    public IntArrayTag getAsIntArray() {
        try {
            return ((IntArrayTag) tag);
        } catch (ClassCastException exception) {
            throw new NBTException("Value is not a int array");
        }
    }

    public short getAsShort() {
        try {
            return ((ShortTag) tag).value();
        } catch (ClassCastException exception) {
            throw new NBTException("Value is not a short");
        }
    }

    public long getAsLong() {
        try {
            return ((LongTag) tag).value();
        } catch (ClassCastException exception) {
            throw new NBTException("Value is not a long");
        }
    }

    public int getAsInt() {
        try {
            return ((IntTag) tag).value();
        } catch (ClassCastException exception) {
            throw new NBTException("Value is not a int");
        }
    }

    public byte getAsByte() {
        try {
            return ((ByteTag) tag).value();
        } catch (ClassCastException exception) {
            throw new NBTException("Value is not a byte");
        }
    }

    public double getAsDouble() {
        try {
            return ((DoubleTag) tag).value();
        } catch (ClassCastException exception) {
            throw new NBTException("Value is not a double");
        }
    }

    public float getAsFloat() {
        try {
            return ((FloatTag) tag).value();
        } catch (ClassCastException exception) {
            throw new NBTException("Value is not a float");
        }
    }

    public String getAsString() {
        try {
            return ((StringTag) tag).value();
        } catch (ClassCastException exception) {
            throw new NBTException("Value is not a string");
        }
    }

    public ITag asTag() {
        return tag;
    }
}

File: src/main/kotlin/io/github/xiefrish2021/compound/NBTElement.kt
package io.github.xiefrish2021.compound

import io.github.xiefrish2021.api.ITag
import io.github.xiefrish2021.api.List
import io.github.xiefrish2021.array.ByteArrayTag
import io.github.xiefrish2021.array.IntArrayTag
import io.github.xiefrish2021.array.LongArrayTag
import io.github.xiefrish2021.list.ListTag
import io.github.xiefrish2021.primitive.PrimitiveTag
import io.github.xiefrish2021.primitive.StringTag
import io.github.xiefrish2021.primitive.number.*

@Suppress("unused")
class NBTElement(val tag: ITag) {
    fun asPrimitive(): PrimitiveTag<*> {
        if (tag is PrimitiveTag<*>) {
            return tag
        }

        return null!!
    }

    fun asString(): String {
        if (tag is StringTag) {
            return tag.value()
        }

        return null!!
    }

    fun asInt(): Int {
        if (tag is IntTag) {
            return tag.value()
        }

        return null!!
    }

    fun asFloat(): Float {
        if (tag is FloatTag) {
            return tag.value()
        }

        return null!!
    }

    fun asDouble(): Double {
        if (tag is DoubleTag) {
            return tag.value()
        }

        return null!!
    }

    fun asByte(): Byte {
        if (tag is ByteTag) {
            return tag.value()
        }

        return null!!
    }

    fun asLong(): Long {
        if (tag is LongTag) {
            return tag.value()
        }

        return null!!
    }

    fun asShort(): Short {
        if (tag is ShortTag) {
            return tag.value()
        }

        return null!!
    }

    fun asList(): List<*> {
        if (tag is ListTag<*>) {
            return tag
        }

        return null!!
    }

    fun asByteArray(): ByteArrayTag {
        if (tag is ByteArrayTag) {
            return tag
        }

        return null!!
    }

    fun asIntArray(): IntArrayTag {
        if (tag is IntArrayTag) {
            return tag
        }

        return null!!
    }

    fun asLongArray(): LongArrayTag {
        if (tag is LongArrayTag) {
            return tag
        }

        return null!!
    }

    fun asCompound(): CompoundTag {
        if (tag is CompoundTag) {
            return tag
        }

        return null!!
    }

    fun asTag(): ITag {
        return tag
    }
}
